jul-ecs-formatter/src/main/java/co/elastic/logging/jul/EcsFormatter.java

```diff
@@ -0,0 +1,103 @@
+/*-
+ * #%L
+ * Java ECS logging
+ * %%
+ * Copyright (C) 2019 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.logging.jul;
+
+import java.util.logging.Formatter;
+import java.util.logging.LogManager;
+import java.util.logging.LogRecord;
+
+import co.elastic.logging.EcsJsonSerializer;
+
+public class EcsFormatter extends Formatter {
+
+    private boolean stackTraceAsArray;
+    private String serviceName;
+    private boolean includeOrigin;
+    private String eventDataset;
+
+    /**
+     * Default constructor. Will read configuration from LogManager properties.
+     */
+    public EcsFormatter() {
+        serviceName = getProperty("co.elastic.logging.jul.EcsFormatter.serviceName", null);
+        includeOrigin = Boolean.getBoolean(getProperty("co.elastic.logging.jul.EcsFormatter.includeOrigin", "false"));
+        stackTraceAsArray = Boolean
+                .getBoolean(getProperty("co.elastic.logging.jul.EcsFormatter.stackTraceAsArray", "false"));
+        eventDataset = getProperty("co.elastic.logging.jul.EcsFormatter.eventDataset", null);
+        eventDataset = EcsJsonSerializer.computeEventDataset(eventDataset, serviceName);
+    }
+
+    @Override
+    public String format(final LogRecord record) {
+        final StringBuilder builder = new StringBuilder();
+        EcsJsonSerializer.serializeObjectStart(builder, record.getMillis());
+        EcsJsonSerializer.serializeLogLevel(builder, record.getLevel().getName());
+        EcsJsonSerializer.serializeFormattedMessage(builder, record.getMessage());
+        EcsJsonSerializer.serializeServiceName(builder, serviceName);
+        EcsJsonSerializer.serializeEventDataset(builder, eventDataset);
+        EcsJsonSerializer.serializeThreadName(builder, "thread-" + record.getThreadID());
+        EcsJsonSerializer.serializeThreadId(builder, record.getThreadID());
+        EcsJsonSerializer.serializeLoggerName(builder, record.getSourceClassName());
+        if (includeOrigin && record.getSourceClassName() != null && record.getSourceMethodName() != null) {
+            EcsJsonSerializer.serializeOrigin(builder, record.getSourceClassName().replaceAll(".*\\.", "") + ".java",
```

Comment:
Use this to get the file name from the class name:
```java
    public void appendFileName(StringBuilder builder, String className) {
        if (className != null) {
            int fileNameEnd = className.indexOf('$');
            if (fileNameEnd < 0) {
                fileNameEnd = className.length();
            }
            int classNameStart = className.lastIndexOf('.');
            if (classNameStart < fileNameEnd && fileNameEnd <= className.length()) {
                JsonUtils.quoteAsString(className, classNameStart + 1, fileNameEnd, builder);
                builder.append(".java");
            } else {
                builder.append("<Unknown>");
            }
        } else {
            builder.append("<Unknown>");
        }
    }
```

You'll have create an overloaded method for `quoteAsString` that takes the range as an argument:

```java
public static void quoteAsString(CharSequence content, int offset, int length, StringBuilder sb)
```

This is what we do in the Elastic APM Java agent too (`co.elastic.apm.agent.impl.transaction.StackFrame#appendFileName`)

Reply:
Done with some corrections. In which case do we need `&& fileNameEnd <= className.length()` condition?

Reply:
That condition indeed seems unnecessary